native_client/tflitemodelstate.cc

```diff
@@ -0,0 +1,266 @@
+  auto erase_begin = std::remove_if(orig_plan.begin(), orig_plan.end(), [&mfcc_plan](int elem) {
+    return std::find(mfcc_plan.begin(), mfcc_plan.end(), elem) != mfcc_plan.end();
+  });
+  orig_plan.erase(erase_begin, orig_plan.end());
+
+  acoustic_exec_plan_ = std::move(orig_plan);
+  mfcc_exec_plan_ = std::move(mfcc_plan);
+
+  TfLiteIntArray* dims_input_node = interpreter_->tensor(input_node_idx_)->dims;
+
+  n_steps_ = dims_input_node->data[1];
+  n_context_ = (dims_input_node->data[2] - 1) / 2;
+  n_features_ = dims_input_node->data[3];
+  mfcc_feats_per_timestep_ = dims_input_node->data[2] * dims_input_node->data[3];
+
+  TfLiteIntArray* dims_logits = interpreter_->tensor(logits_idx_)->dims;
+  const int final_dim_size = dims_logits->data[1] - 1;
+  if (final_dim_size != alphabet_->GetSize()) {
+    std::cerr << "Error: Alphabet size does not match loaded model: alphabet "
+              << "has size " << alphabet_->GetSize()
+              << ", but model has " << final_dim_size
+              << " classes in its output. Make sure you're passing an alphabet "
+              << "file with the same size as the one used for training."
+              << std::endl;
+    return DS_ERR_INVALID_ALPHABET;
+  }
+
+  TfLiteIntArray* dims_c = interpreter_->tensor(previous_state_c_idx_)->dims;
+  TfLiteIntArray* dims_h = interpreter_->tensor(previous_state_h_idx_)->dims;
+  assert(dims_c->data[1] == dims_h->data[1]);
+  assert(state_size_ > 0);
+  state_size_ = dims_c->data[1];
+
+  return DS_ERR_OK;
+}
+
+void
+TFLiteModelState::copy_vector_to_tensor(const vector<float>& vec,
+                                        int tensor_idx,
+                                        int num_elements)
+{
+  float* tensor = interpreter_->typed_tensor<float>(tensor_idx);
+  int i;
+  for (i = 0; i < vec.size(); ++i) {
+    tensor[i] = vec[i];
+  }
+  for (; i < num_elements; ++i) {
+    tensor[i] = 0.f;
+  }
+}
+
+void
+TFLiteModelState::copy_tensor_to_vector(int tensor_idx,
+                                        int num_elements,
+                                        vector<float>& vec)
+{
+  float* tensor = interpreter_->typed_tensor<float>(tensor_idx);
+  for (int i = 0; i < num_elements; ++i) {
+    vec.push_back(tensor[i]);
+  }
+}
+
+void
+TFLiteModelState::infer(const vector<float>& mfcc,
+                        unsigned int n_frames,
+                        const vector<float>& previous_state_c,
+                        const vector<float>& previous_state_h,
+                        vector<float>& logits_output,
+                        vector<float>& state_c_output,
+                        vector<float>& state_h_output)
+{
+  const size_t num_classes = alphabet_->GetSize() + 1; // +1 for blank
+
+  // Feeding input_node
+  copy_vector_to_tensor(mfcc, input_node_idx_, n_frames*mfcc_feats_per_timestep_);
+
+  // Feeding previous_state_c, previous_state_h
+  assert(previous_state_c.size() == state_size_);
```

Comment:
maybe assert within the copy functions ?

Reply:
The copy functions are agnostic to what tensor you're copying stuff into, and they're meant to work with different size vector and `num_elements`, so I can't put these asserts there.